### PR TITLE
Feat: NOR unit tests

### PR DESCRIPTION
**Pull request by TheDZhon**
### NodeOperatorsRegistry unit tests

- [x] Old mock moved to `NodeOperatorsRegistry__Harness`
- [x] New helpers: `lib/nor.ts`

#### NodeOperatorsRegistry:auxiliary
- [x] `unsafeUpdateValidatorsCount`
- [x] `updateTargetValidatorsLimits`
- [x] `onWithdrawalCredentialsChanged`
- [x] `invalidateReadyToDepositKeysRange`
- [x] `getRewardsDistribution`

#### NodeOperatorsRegistry:initialize-and-upgrade
- [x] `initialize`
- [x] `finalizeUpgrade_v2`

#### NodeOperatorsRegistry:management
- [x] `addNodeOperator`
- [x] `activateNodeOperator`
- [x] `deactivateNodeOperator`
- [x] `setNodeOperatorName`
- [x] `setNodeOperatorRewardAddress`
- [x] `setNodeOperatorStakingLimit`
- [x] `getNodeOperator`
- [x] `getType`
- [x] `getStakingModuleSummary`
- [x] `getNodeOperatorSummary`
- [x] `getNodeOperatorsCount`
- [x] `getActiveNodeOperatorsCount`
- [x] `getNodeOperatorIsActive`
- [x] `getNodeOperatorIds`
- [x] `getNonce`
- [x] `getKeysOpIndex`
- [x] `getLocator`
- [x] `getStuckPenaltyDelay`
- [x] `setStuckPenaltyDelay`

#### NodeOperatorsRegistry:rewards-penalties

- [x] `onRewardsMinted`
- [x] `updateStuckValidatorsCount`
- [x] `updateExitedValidatorsCount`
- [x] `updateRefundedValidatorsCount`
- [x] `onExitedAndStuckValidatorsCountsUpdated`
- [x] `isOperatorPenalized`
- [x] `isOperatorPenaltyCleared`
- [x] `clearNodeOperatorPenalty`

#### NodeOperatorsRegistry:signing-keys

- [x] `obtainDepositData`
- [x] `addSigningKeys`
- [x] `addSigningKeysOperatorBH`
- [x] `removeSigningKey`
- [x] `removeSigningKeys`
- [x] `removeSigningKeyOperatorBH`
- [x] `removeSigningKeysOperatorBH`
- [x] `getTotalSigningKeyCount`
- [x] `getUnusedSigningKeyCount`
- [x] `getSigningKey`
- [x] `getSigningKeys`

### CODE REVIEW

**Comment by TheDZhon** [comment]
image

File: lib/nor.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  };
+}
+
+/***
+ * Returns a hexadecimal representation of the given number
+ * @param {bigint} n number to represent in hex
+ * @param {number} byteLen bytes to pad
+ * @returns {string} hexadecimal string
+ */
+function _hex(n: bigint, byteLen?: number): string {
```
Comment (tamtamchik) [inline lib/nor.ts:142]: It is worth moving to `string.ts` and removing `bigintToHex` along with the `bigint-conversion` dependency. Wdyt?

**Review by tamtamchik — COMMENTED** [review]
Terrific! Great work! 🙇

I highlighted some uncovered branches in the NOR contract to get some early feedback; I will continue with the PR review a bit later this week

File: lib/nor.ts [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+ * @param {bigint} config.refundedValidatorsKeysCount Repaid keys count of the new node operator
+ * @param {bigint} config.isActive The active state of new node operator
+ * @returns {bigint} newOperatorId Id of newly added Node Operator
+ */
+export async function addNodeOperator(
+  norMock: NodeOperatorsRegistry__Harness,
+  norManager: HardhatEthersSigner,
+  config: NodeOperatorConfig,
+): Promise<bigint> {
+  const isActive = config.isActive === undefined ? true : config.isActive;
```
[the remark this answers is not in the document]
Comment (TheDZhon) [reply lib/nor.ts:43]: fixed

File: lib/nor.ts (hunk 1 above)
Comment (TheDZhon) [reply lib/nor.ts]: yeah, good suggestion, fixed :+1:

File: test/0.4.24/nor/nor.rewards.penalties.flow.test.ts [hunk 3]
```diff
@@ -0,0 +1,752 @@
...
+    it("Reverts if has no STAKING_ROUTER_ROLE assigned", async () => {
+      await expect(nor.onRewardsMinted(10n)).to.be.revertedWith("APP_AUTH_FAILED");
+    });
+
+    it("Does nothing yet if called by StakingRouter", async () => {
+      await nor.connect(stakingRouter).onRewardsMinted(10n);
+    });
+  });
+
+  context("updateStuckValidatorsCount", () => {
```
[the remark this answers is not in the document]
Comment (TheDZhon) [reply test/0.4.24/nor/nor.rewards.penalties.flow.test.ts:164]: added

File: test/0.4.24/nor/nor.aux.test.ts [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+            NODE_OPERATORS[secondNodeOperatorId].depositedSigningKeysCount,
+        )
+        .to.emit(nor, "KeysOpIndexSet")
+        .withArgs(nonce + 1n)
+        .to.emit(nor, "NonceChanged")
+        .withArgs(nonce + 1n);
+    });
+  });
+
+  context("invalidateReadyToDepositKeysRange", () => {
```
[the remark this answers is not in the document]
Comment (TheDZhon) [reply test/0.4.24/nor/nor.aux.test.ts:353]: fixed

File: test/0.4.24/nor/nor.rewards.penalties.flow.test.ts [hunk 5]
```diff
@@ -0,0 +1,752 @@
...
+
+      await expect(nor.connect(stakingRouter).updateRefundedValidatorsCount(firstNodeOperatorId, 0n))
+        .to.emit(nor, "StuckPenaltyStateChanged")
+        .withArgs(firstNodeOperatorId, 0n, 0n, 0n)
+        .to.not.emit(nor, "KeysOpIndexSet")
+        .to.not.emit(nor, "NonceChanged");
+    });
+  });
+
+  context("onExitedAndStuckValidatorsCountsUpdated", () => {
```
[the remark this answers is not in the document]
Comment (TheDZhon) [reply test/0.4.24/nor/nor.rewards.penalties.flow.test.ts:447]: fixed

File: test/0.4.24/nor/nor.signing.keys.test.ts [hunk 6]
```diff
@@ -0,0 +1,1082 @@
...
+    expect(await addNodeOperator(nor, nodeOperatorsManager, NODE_OPERATORS[secondNodeOperatorId])).to.be.equal(
+      secondNodeOperatorId,
+    );
+  });
+
+  beforeEach(async () => (originalState = await Snapshot.take()));
+
+  afterEach(async () => await Snapshot.restore(originalState));
+
+  context("obtainDepositData", () => {
```
[the remark this answers is not in the document]
Comment (TheDZhon) [reply test/0.4.24/nor/nor.signing.keys.test.ts:177]: the second part is fixed

[the remark this answers is not in the document]
Comment (TheDZhon) [reply test/0.4.24/nor/nor.signing.keys.test.ts:177]: Seems the condition `if (depositedSigningKeysCountAfter == depositedSigningKeysCountBefore) continue` is unreachable in the wild, will figure out the plan tomorrow in the morning

**Review by mymphe — APPROVED** [review]
Absolutely titanic work! All tests are uniformly written in accordance with the established best practices! Thank you!

**Review by tamtamchik — APPROVED** [review]
🦄 AWESOME WORK!

**Comment by TheDZhon** [comment]
image

**Comment by TheDZhon** [comment]
image

File: test/0.4.24/nor/nor.signing.keys.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (TheDZhon) [reply test/0.4.24/nor/nor.signing.keys.test.ts:177]: fixed

**Pull request merged**